### ISSUE TITLE
CRASH REPORT after connecting from the other side of the globe

### ISSUE THREAD

**Issue by BertRobbenSony** [erlio/vernemq#41]
Hi guys,

Not sure if this is the right forum, but I'll go for it anyway.

I'm using VerneMQ here as an mqtt server for a prototype application and that works fine (connection/publish/subscribe). However, when my colleagues from Japan try to do the same, they fail to connect to my server. If I then look in the logs I see the following CRASH REPORTs (they tried to connect twice).

2015-08-28 12:18:33.434 [error] <0.2751.9> CRASH REPORT Process <0.2751.9> with 0 neighbours crashed with reason: no match of right hand value {error,enotconn} in vmq_ranch:init/4 line 66
2015-08-28 12:18:33.435 [error] <0.270.0> Ranch listener {{192,168,116,112},1883} had connection process started with vmq_ranch:start_link/4 at <0.2751.9> exit with reason: {badmatch,{error,enotconn}}

2015-08-28 12:18:55.087 [error] <0.2791.9> CRASH REPORT Process <0.2791.9> with 0 neighbours crashed with reason: no match of right hand value {error,enotconn} in vmq_ranch:init/4 line 66
2015-08-28 12:18:55.087 [error] <0.270.0> Ranch listener {{192,168,116,112},1883} had connection process started with vmq_ranch:start_link/4 at <0.2791.9> exit with reason: {badmatch,{error,enotconn}}

Am I doing something wrong or is this a server issue?

Here are some more details:

I'm running Vernemq 0.10.1-ef5a45b3 on Ubuntu (64 bit).
Client is an android phone using the Eclipse PAHO mqtt client.
Network setup is client => internet => reverse proxy on AWS => ssh tunnel => vernemq server. From my location (Europe) this works fine, but I can imagine different delays when calling from Japan.

cheers,

Bert

**Comment by dergraf** [comment]
Hi Bert,

thanks for your report! The error you're seeing is triggered right after VerneMQ accepted the new TCP connection. At this point VerneMQ wants to figure out the origin IP address/port which causes this error (logged in an ugly way, we'll improve that). 
However, this error should actually only happen if the connection was closed in the meantime (for whatever reasons). We've seen similar errors where a mobile network provider forcefully closed TCP connections that e.g. don't transport HTTP. Enforcing to use SSL helped in such cases. 

Are your colleagues connecting through a mobile network?

Best,
Andre

**Comment by BertRobbenSony** [comment]
Hi Andre,

thanks for the quick reply! I'll check with my colleagues, but I'm almost sure they are using WIFI. I'll ask them to switch to ssl to see if that would help.

thanks,

Bert

**Comment by BertRobbenSony** [comment]
There was indeed a networking issue. This has been fixed now and clients can connect as they should.

**Comment by dergraf** [comment]
Cool! Can you tell us more about the networking issue?

**Comment by BertRobbenSony** [comment]
Hi Andre,

It’s not 100% clear to me what exactly happened, as more than one thing was going on at the same time. This is what I know:


(1)    At the same time the crash reports were generated, I was doing some load testing. So the load on the node at that time was high. Vernemq in particular was using quite some CPU. In addition another process running on the same host as the mqtt server was going out-of-memory. So resources were scarce.

(2)    The app that was making the mqtt connection was not properly displaying the status to the user. So my colleagues couldn’t really see what was going on. So they probably switched the app on and off all the time (this will cause a disconnect to the current mqtt connection and creation of a new one).

a.       Because of this display issue, I’m now even no longer sure that there was a real issue in setting up the mqtt connection. The CRASH REPORTs I saw could have been caused by my colleagues themselves hitting the “stop” button before the connection was properly initialized.

(3)    My colleagues also turned wifi on and off. I don’t know why they did this, I just saw it from their logs. The app only uses mqtt when there is an active wifi connection. In case there is no wifi it will disconnect.

(4)    After I fixed the out-of-memory issue and stopped my load testing, I tried to connect myself from Japan (using a VPN). That always worked as it should. So I could not reproduce what they did.

As a side note, I noticed that the cpu usage of the mqtt server was quite high (during my initial load testing) when I connected mqtt-spy to it and subscribing to virtually all messages. I do this from time to time as a kind of debugging help to see that all my messages are properly published. When I unsubscribe mqtt-spy, cpu load falls back to an expected value. I don’t really know why this happens, but it’s not really an issue for me as mqtt-spy is just a tool that I use during development so during real runs it is not used.

Bert
